module/profiler/profiler.go

```diff
@@ -143,11 +146,96 @@ func newProfileFunc(name string) profileFunc {
 	}
 }
 
-func (p *AutoProfiler) pprofAllocs(w io.Writer) error {
+func (p *AutoProfiler) goHeapProfile() (*profile.Profile, error) {
 	// Forces the GC before taking each of the heap profiles and improves the profile accuracy.
 	// Autoprofiler runs very infrequently so performance impact is minimal.
 	runtime.GC()
-	return newProfileFunc("allocs")(w)
+
+	buf := &bytes.Buffer{}
+	err := newProfileFunc("heap")(buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate allocs profile: %w", err)
+	}
+
+	prof, err := profile.Parse(buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse allocs profile: %w", err)
+	}
+	prof.TimeNanos = time.Now().UnixNano()
+
+	if got := len(prof.SampleType); got != 4 {
```

Comment:
These checks seems very brittle.

1. what happens if the profile returns more types than the expected ones?
2. this assumes a specific sample type ordering.  what happens if the profile returns the samples in different order?

Reply:
Your coding standards are [higher than Google's](https://github.com/googleapis/google-cloud-go/blob/main/profiler/heap.go#L107), but fair enough, this code is indeed brittle.

Let me rewrite it in a more robust way, but for that I'll need to improve testing quite a bit.

Reply:
real googler backend people write c++ =P